### PR TITLE
allow the age warning to ignore on hold jobs with certain words

### PR DESCRIPTION
**Pull request by rwilcox**
We want to support not warning users for every on hold job that's going to be cancelled.

Now, the log job canceller allows us to pass an ignore parameter, and workflows that are paused we examine all the on hold jobs to see if all those jobs contain the ignore word (or words). If so, do not warn the user.

### CODE REVIEW

File: src/commands/circleci-continue-long-job-cancel.yml [hunk 1]
```diff
@@ -11,14 +11,18 @@ parameters:
     description: the name of the environmental variable that contains the circleci token
     default: "CIRCLECI_TOKEN"
     type: string
+  ignore-jobs-with-name:
+    description: do not warn people about jobs with this name
```
Comment (AlexanderMann) [inline src/commands/circleci-continue-long-job-cancel.yml:15]: the text of this description is slightly different at every level. It looks like the python says it filters on it being a substring, whereas this looks like an exact match

File: src/scripts/circleci-job-canceller/Modules/circle_utils.py [hunk 2]
```diff
@@ -8,6 +8,10 @@ def get_all_items(relative_url, headers, pagination_limit=5):
     pages_iterated = 0
     while True:
         res = requests.get(temp_url, headers=headers).json()
+        if (res.get("message") == 'An internal server error occurred.'):
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/Modules/circle_utils.py:11]: Does this not get caught below in the try catch? Maybe extra context there would suffice?

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 3]
```diff
@@ -5,6 +5,8 @@
 import datetime
 import sys
 import os.path
+import pprint
+import itertools
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:9]: alphabets

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 4]
```diff
@@ -30,6 +32,23 @@ def get_workflow_started_by(current_workflow, headers):
...
 
+def get_workflow_pending_approval_jobs(workflow_id, headers):
+    """
+    the top object of the workflow only tells us that the workflow is on hold, not why.
+    Iterate through all the jobs of the workflow checking in particular for approval jobs
+    and return information about approval jobs in this workflow that are on hold
+    (so then we can filter them later)
+    """
+    url = f"https://circleci.com/api/v2/workflow/{workflow_id}/job"
+    workflow_info = requests.get(url, headers=headers).json()
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:43]: This'll need pagination for sure

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 5]
```diff
@@ -30,6 +32,23 @@ def get_workflow_started_by(current_workflow, headers):
...
+
+    all_jobs = workflow_info.get("items")
+    output = []
+
+    [ output.append(x) for x in all_jobs if (x.get("type") == "approval") and (x.get("status") == "on_hold") ]
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:48]: Imo, I'd just nix `output` altogether and just return the list comprehension

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 6]
```diff
@@ -60,7 +79,23 @@ def find_old_workflow_ids(repo_slug, window_start, window_end, headers):
                     yield {"job_status": "too_old", "name": current_workflow['name'], "id": current_workflow['id'], "username": username}
 
 
-def main(circleapitoken, orgreposlug, output_file, commit):
+def do_we_care_about_this_pipeline(current_info, ignore, headers):
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:82]: I think `filter_ignored_jobs` would be a clearer name, and then you can simply use empty list checking in the `main` to decide whether to continue with a workflow.

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 7]
```diff
@@ -30,6 +32,23 @@ def get_workflow_started_by(current_workflow, headers):
     return username
 
 
+def get_workflow_pending_approval_jobs(workflow_id, headers):
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:35]: If we passed in the `ignore` parameter here, we could even paginate through the responses, and upon finding _any_ job which is pending but not named incorrectly, we could terminate 🤷

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 8]
```diff
@@ -60,7 +79,23 @@ def find_old_workflow_ids(repo_slug, window_start, window_end, headers):
...
-def main(circleapitoken, orgreposlug, output_file, commit):
+def do_we_care_about_this_pipeline(current_info, ignore, headers):
+    if ignore == "":
+        return True
+
+    pending_approvals = get_workflow_pending_approval_jobs(current_info['id'], headers)
+
+    # we want to know if all of the names of the approval jobs waiting contain the ignore keyword
+    # if all of them do, then we do NOT care about this pipeline
+    not_ignored_jobs = itertools.dropwhile( lambda x: x['name'].find(ignore) > -1 , pending_approvals )
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:90]: can flip this to say whether _any_ job exists which _doesn't_ match the `ignore`

**Review by AlexanderMann — CHANGES_REQUESTED** [review]

File: src/scripts/circleci-job-canceller/Modules/circle_utils.py (hunk 2 above)
Comment (rwilcox) [reply src/scripts/circleci-job-canceller/Modules/circle_utils.py:11]: so yes... and it gets caught and rethrown, and the exception thrown exits the program.

I don't know if it's a good idea to squash the error - forcing the pagination iteration to be over - or figure out why exactly in might be sad (or maybe it was just cranky yesterday afternoon) or write things in a more streaming fashion so an error won't throw away progress

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 9]
```diff
@@ -114,8 +154,9 @@ def main(circleapitoken, orgreposlug, n_windows, output_file, commit):
                         help="Number of windows to look back across. Default window length is 2 hours.",
                         type=int,
                         default=6)
+    parser.add_argument("--ignore", help="if all awaiting approval jobs in the pipeline contain this word, do not age warn about it. Multiple word supported by delimiting with ,", default="")
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:157]: An explanation of what the logic is when there are multiple would be 👍

File: src/scripts/circleci-job-canceller/long_job_canceller.py [hunk 10]
```diff
@@ -68,7 +85,26 @@ def find_old_workflow_ids(repo_slug, window_start, window_end, headers):
...
+    for current in list:
+        if ( str.find(current) > -1 ):
+            return True
+    return False
+
+
+def has_only_ignored_jobs(current_info, ignore, headers):
+    pending_approvals = get_workflow_pending_approval_jobs(current_info['id'], headers)
+
+    for current in itertools.filterfalse(
```
Comment (AlexanderMann) [inline src/scripts/circleci-job-canceller/long_job_canceller.py:98]: Any reason to have the `itertools` + `lambda` here? Seems harder/less Pythonic to read than a:

```py
for current in get_workflow_pending_approval_jobs(current_info['id'], headers):
  if not matches_any_in_list(current['name], ignore):
    return False
return True
```

File: src/commands/circleci-continue-long-job-cancel.yml [hunk 11]
```diff
@@ -11,6 +11,10 @@ parameters:
     description: the name of the environmental variable that contains the circleci token
     default: "CIRCLECI_TOKEN"
     type: string
+  ignore-jobs-containing:
+    description: do not warn people about workflows where all the on hold job names contain this string. Multiple strings seperated by comma, eg development,delete
```
Comment (AlexanderMann) [inline src/commands/circleci-continue-long-job-cancel.yml:15]:
```suggestion
    description: Do not warn about workflows where all of the on hold job names contain a string from this comma separated list
```

fwiw, I'd update all of the args to match, and the descriptions

**Review by AlexanderMann — CHANGES_REQUESTED** [review]

File: src/commands/circleci-continue-long-job-cancel.yml [hunk 12]
```diff
@@ -21,7 +25,7 @@ steps:
       working_directory: /tmp/apollo_internal_platform_orb/src/scripts/circleci-job-canceller
       command: |
         python3 -m pip install -r requirements.txt
-        python3 long_job_canceller.py $<< parameters.circleci-token-variable >> << parameters.org-repo-slug >> --commit --n-windows << parameters.n-windows >>
+        python3 long_job_canceller.py $<< parameters.circleci-token-variable >> << parameters.org-repo-slug >> --commit --n-windows << parameters.n-windows >> --ignore=<< parameters.ignore-jobs-containing >>
```
Comment (AlexanderMann) [inline src/commands/circleci-continue-long-job-cancel.yml:28]: same nit about the chance for all of these variable/argument names to be able to match. Less cognitive overheard overall

**Review by AlexanderMann — CHANGES_REQUESTED** [review]

**Review by AlexanderMann — CHANGES_REQUESTED** [review]

**Pull request merged**